Zend/zend_fibers.c

```diff
@@ -0,0 +1,778 @@
+				zend_clear_exception();
+			} else {
+				zend_exception_error(EG(exception), E_ERROR);
+			}
+		} else {
+			fiber->status = ZEND_FIBER_STATUS_THREW;
+		}
+	} else {
+		fiber->status = ZEND_FIBER_STATUS_RETURNED;
+	}
+
+	zend_vm_stack_destroy();
+	fiber->execute_data = NULL;
+
+	// Remove reference added at last resume.
+	GC_DELREF(&fiber->std);
+}
+/* }}} */
+
+static zend_object *zend_fiber_object_create(zend_class_entry *ce) /* {{{ */
+{
+	zend_fiber *fiber;
+
+	fiber = emalloc(sizeof(zend_fiber));
+	ZEND_SECURE_ZERO(fiber, sizeof(zend_fiber));
+
+	fiber->id = EG(next_fiber_id)++;
+
+	zend_object_std_init(&fiber->std, ce);
+	fiber->std.handlers = &zend_fiber_handlers;
+
+	ZVAL_UNDEF(&fiber->value);
+
+	zend_hash_index_add_ptr(&EG(fibers), fiber->id, fiber);
+
+	return &fiber->std;
+}
+/* }}} */
+
+static void zend_fiber_object_destroy(zend_object *object) /* {{{ */
+{
+	zend_fiber *fiber = (zend_fiber *) object;
+
+	if (UNEXPECTED(fiber->status == ZEND_FIBER_STATUS_INIT)) {
+		// Fiber was never started, so we need to release the reference to the callback.
+		zval_ptr_dtor(&fiber->fci.function_name);
+	} else {
+		if (fiber->status == ZEND_FIBER_STATUS_SUSPENDED) {
+			zend_object *exception = EG(exception);
+			EG(exception) = NULL;
+			fiber->status = ZEND_FIBER_STATUS_SHUTDOWN;
+			zend_fiber_switch_to(fiber);
+			EG(exception) = exception;
+		}
+
+		zval_ptr_dtor(&fiber->value);
+	}
+
+	zend_hash_index_del(&EG(fibers), fiber->id);
+
+	zend_fiber_destroy_context(&fiber->context);
+
+	zend_object_std_dtor(&fiber->std);
+}
+/* }}} */
+
+void zend_fiber_cleanup(void) /* {{{ */
+{
+	zend_fiber *fiber;
+	zend_object *exception = EG(exception);
+	EG(exception) = NULL;
+
+	ZEND_HASH_REVERSE_FOREACH_PTR(&EG(fibers), fiber) {
+		if (fiber->status == ZEND_FIBER_STATUS_SUSPENDED) {
+			fiber->status = ZEND_FIBER_STATUS_SHUTDOWN;
+			GC_ADDREF(&fiber->std);
+			zend_fiber_switch_to(fiber);
+		}
+	} ZEND_HASH_FOREACH_END();
+
+	EG(exception) = exception;
+}
+/* }}} */
+
+/* {{{ proto Fiber::__construct(callable $callback) */
+ZEND_METHOD(Fiber, __construct)
+{
+	zend_fiber *fiber = (zend_fiber *) Z_OBJ_P(getThis());
+
+	ZEND_PARSE_PARAMETERS_START(1, 1)
+		Z_PARAM_FUNC_EX(fiber->fci, fiber->fci_cache, 0, 0)
+	ZEND_PARSE_PARAMETERS_END();
+
+	// Keep a reference to closures or callable objects until the fiber is started.
+	Z_TRY_ADDREF(fiber->fci.function_name);
+}
+/* }}} */
+
+/* {{{ proto mixed Fiber::start(mixed ...$args) */
+ZEND_METHOD(Fiber, start)
+{
+	zend_fiber *fiber = (zend_fiber *) Z_OBJ_P(getThis());
+
+	if (fiber->status != ZEND_FIBER_STATUS_INIT) {
+		zend_throw_error(zend_ce_fiber_error, "Cannot start a fiber that has already been started");
+		return;
+	}
+
+	fiber->fci.params = ZEND_CALL_ARG(execute_data, 1);
+	fiber->fci.param_count = ZEND_CALL_NUM_ARGS(execute_data);
+
+	if (!zend_fiber_init_context(&fiber->context, zend_fiber_execute, EG(fiber_stack_size))) {
+		zend_throw_error(NULL, "Could not create fiber context");
```

Comment:
`FiberError` was reserved for using the fiber API incorrectly, this error is out of the user's control, so I felt `Error` was more appropriate. Does that sound fine or should I use `FiberError` or something else?

Reply:
Hm yeah, this is tricky. If we're being pedantic, this should throw an `Exception`, because this is an environment failure condition and does not (always) indicate programming error. It's something an application probably needs to handle, in case too many fibers are created we run out of memory.

... Actually, it seems really easy to trigger this. I'm using this script to me:
```
$fibers = [];
while (true) {
    echo $i++, "\n";
    $fiber = new Fiber(function() { Fiber::suspend(); });
    $fiber->start();
    $fibers[] = $fiber;
}
```
Regardless of used fiber.stack_size, this fails at about 32507 fibers. The actually failing call is the mprotect on the guard pages (this is for 128k):

```
mmap(NULL, 135168, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_STACK, -1, 0) = 0x7f32fcbe2000
mprotect(0x7f32fcbe2000, 4096, PROT_NONE) = -1 ENOMEM (Cannot allocate memory)
```

Any idea why that is?

Reply:
Check out [this issue](https://github.com/amphp/ext-fiber/issues/11#issuecomment-791864364), I think it has to do with a per-process limit on the number of mapped memory regions.

Reply:
Apparently this is because of `vm.max_map_count = 65530`.

Reply:
> If we're being pedantic, this should throw an Exception, because this is an environment failure condition and does not (always) indicate programming error.

I agree, I'll change it to `Exception`.